plantcv/plantcv/transform/warp.py

```diff
@@ -3,56 +3,87 @@
-    pts = 4 coordinates on img1
-    refpts = 4 coordinates on img2
-    method = method of finding the transformation. 'default', 'ransac', 'lmeds', 'rho'
+    img = image to be warped
+    refimg = image to be used as reference
+    pts = 4 coordinates on img
+    refpts = 4 coordinates on refimg
+    method = Available options are 'default', 'ransac', 'lmeds', 'rho' which correspond to the opencv methods
+    and vary based on whether they handle outlier points and how they handle outlier points.
+      - 'default': a regular method (the least squares method) using all the points
+      - 'ransac': RANSAC-based robust method
+        ([Random sample consensus](https://en.wikipedia.org/wiki/Random_sample_consensus#:~:text=An%20advantage%20of%20RANSAC%20is,present%20in%20the%20data%20set.&text=RANSAC%20can%20only%20estimate%20one%20model%20for%20a%20particular%20data%20set.))
+      - 'lmeds': Least-Median robust method
+        ([Least Median of Squares](http://www-sop.inria.fr/odyssee/software/old_robotvis/Tutorial-Estim/node25.html))
+      - 'RHO': PROSAC-based robust method ([PROSAC]
     Returns:
     warped_img = warped image
+    mat = transformation matrix
 
     :param img: numpy.ndarray
     :param refimg: numpy.ndarray
     :param pts: list of tuples
     :param refpts: list of tuples
     :param method: str
+    :return mat: numpy.ndarray
     :return warped_img: numpy.ndarray
+    :return mat: numpy.ndarray
     """
 
-    params.device += 1
-
-    if len(pts) != 4 or len(refpts) != 4:
-        fatal_error('Please provide 4 pairs of corresponding coordinates.')
-    if len(img.shape) > 2:
-        fatal_error('The input `img` should be grayscale or binary.')
+    if len(pts) != len(refpts):
+        fatal_error('Please provide same number of corresponding coordinates.')
+    if not(len(pts) >=4 and len(refpts) >= 4):
+        fatal_error('Please provide at least 4 pairs of points!')
+    # convert coordinates to int if they are not int
+    pts = [tuple(map(int, tup)) for tup in pts]
+    refpts = [tuple(map(int, tup)) for tup in refpts]
 
     methods = {
         'default': 0,
         'ransac': cv2.RANSAC,
         'lmeds': cv2.LMEDS,
         'rho': cv2.RHO}
 
+    shape_img = img.shape
     shape_ref = refimg.shape
     rows_ref, cols_ref = shape_ref[0:2]
+    rows_img, cols_img = shape_img[0:2]
 
     # convert list of tuples to array for cv2 functions
     ptsarr = np.array(pts, dtype='float32')
     refptsarr = np.array(refpts, dtype='float32')
 
     # find tranformation matrix and warp
-    mat, _ = cv2.findHomography(ptsarr, refptsarr, method=methods.get(method))
+    mat, status = cv2.findHomography(ptsarr, refptsarr, method=methods.get(method))
+    if mat is None:
+        fatal_error("Cannot calculate a robust model with given corresponding coordinates and with desired robust estimation algorithm {}!".format(method))
     warped_img = cv2.warpPerspective(src=img, M=mat, dsize=(cols_ref, rows_ref))
```

Comment:
I am not sure about this. 
 @nfahlgren I remember you mentioned that we should avoid using plantcv function in test, but what if in this case?

Reply:
I agree generally about avoiding duplication but I think it's a bit simpler, in this case, to run the OpenCV function directly since there's no other pre-processing going on. If we run `warp_align`, we have to cache the debug state, turn it off, turn it back on after, and run through the preprocessing code in `warp_algin` that has no effect here (I think?).

Reply:
You are right!